### ISSUE TITLE
Feature wishlist for WebExtensions port

### ISSUE THREAD

**Issue by bovine3dom** [vimperator/vimperator-labs#729]
See #705 first. I'd like to use this issue specifically to understand what features people would miss from Vimperator, so that we can try to ensure that any WebExtensions port/successor contains them. What follows is my personal opinion:

# High priority
 - user remappable keys to arbitrary commands (e.g. `bind s ogoogle `)
 - moving around page
 - link hinting mode
 - insert/passthrough mode
 - find in page mode
 - history navigation
 - tab switching / navigation (`:buffer ` style)
 - opening/closing URLs / tabs / windows
 - consistent UX: e.g, works on `about:newtab`, `about:home`, hopefully others.

# Medium priority
 - bookmarks
 - hide Firefox GUI
 - window switching

# Low priority
 - download UI (can't use about:downloads unless Firefox devs change their minds).
 - `:js`
 - `VIMPERATOR_INIT` 

Please disagree with me below 😄.

**Comment by whirm** [comment]
The external editor functionality (CTL+I) Is super useful to me, I use it very often to edit all sorts of stuff in Emacs.

**Comment by cmcaine** [comment]
Emacs!? Ban this sick filth.

**Comment by whirm** [comment]
I use it with Spacemacs/evil if that offends you any less ;)

**Comment by cmcaine** [comment]
Slightly ;P

We can do this with the runtime.nativeConnect API, but users who want to use it will have to install a couple of files for us: a simple shell script and a .json file telling Firefox that we're allowed to talk to that shell script.

On windows we need a registry key, a json file and a batch file, but same principle.

If we get a Filesystem API that lets us write to disk, we might be able to install it automatically on *nix.

**Comment by MoSal** [comment]
* Extended hints.
* `:set gui`
* `:tab*` commands.
* `:style*` commands.
* `:bdelete` (with `arg`)
* `:restart`

**Comment by gkatsev** [comment]
Looks like a decent list but needs to be reconciled with what is possible and not possible with web-ext. For example, hiding the firefox GUI is probably not going to be possible even with the new theming support.

Also, can you elaborate some more about "moving around the page"?

**Comment by bovine3dom** [comment]
I mean hjkl, ^D, ^U etc.

I think it makes sense to work out which features we desire most, and then talk about feasibility after, so we know how hard to fight, and where we can make compromises more easily.

The theming, for example, is something we can lobby for, and if we must, we could make our own WebExtensions API to hide GUI which can be used in the developer edition only.

**Comment by gkatsev** [comment]
Oh, of course, those kind of movement.

We probably should prioritize things by web-ext support as well. So, high things should be core vimperator stuff and those that web-ext supports (well, key bindings are an exception to something we need to have regardless of web-ext support).

**Comment by nachoparker** [comment]
thanks for the list
ctrl+I  is awesome. It is good to see that there is a way
I also use ctrl+t quite a bit. really handy

> The theming, for example, is something we can lobby for, and if we must, we could make our own WebExtensions API to hide GUI which can be used in the developer edition only.

Agree, when the rest is mature, we shouldn't be too scared of changing FF itself. Just small changes close always to upstream.

> We probably should prioritize things by web-ext support as well.

Definitely.. I would focus on what we can do easily first, then start the extreme tweaks ;)

**Comment by cmcaine** [comment]
I think this bug will be most useful to implementers as a list of desired features, without concern for whether they can be implemented or not. **Just tell us what you would miss from Vimperator if it was gone.**

Let implementers worry about filtering by feasibility - feasibility will be a moving target, anyway.

**Comment by nachoparker** [comment]
Some more 

- o and O to open url/search (modifying current in upper case)
- t and T to open a new tab( modifying current in upper case)
- d to close a tab
- p and P to open URL in clipboard in tab (current or newr
- u to bring back a closed tab
- CTRL+D CTRL+U  CTRL+E CTRL+Y
- vimlike jumps (G, gg, '' ...)
- quick marks
- maps keys to javascript code (low priority, but I use it a lot also)
- find (/) with hlsearch support, n and N to navigate results
- CTRL H and CTRL L to navigate history

well, some have already been said

**Comment by KamphuisAlex** [comment]
Those are the important ones I would say.

I would however like to add support for custom search engines, e.g. use yt for searching on youtube just like is possible in the normal address-bar and in vimperator now.

**Comment by gkatsev** [comment]
@cmcaine makes sense. We can then filter things later.

**Comment by MoSal** [comment]
Are we supposed to list everything?

More features:

* `:set complete`
* `:set encoding`
* `:zoom` (with value)

**Comment by nachoparker** [comment]
We are just stating what is more useful for now. Anyway, for completeness I will post here the main sections in Vimperator help. This does not mean that we are doing all this, but just to facilitate people say what they use the most.

- Browsing:        Basic mappings and commands needed for a browsing session (how to open a web page or go back in history).
- Buffer:        Operations on the current document like scrolling or copying text.
- Command-line mode:        Command-line editing.
- Insert mode:     Insert-mode editing.
- Options:    A description of all options.
- Text search commands:  Searching for text in the current buffer.
- Tabs: Managing your tabbed browsing session.
- Hints:  Selecting hyperlinks and other page elements.
- Key mappings, abbreviations, and user-defined commands:
- Defining new key mappings, abbreviations and user commands.
- Expression evaluation: Executing JavaScript.
- Marks: Using bookmarks, QuickMarks, history and local marks.
- Repeating commands: Using macros to repeat recurring workflows.
- Automatic commands:
- Automatically executing code on certain events.
- Printing:  Printing pages.
 - Vimperator's GUI:     Accessing Firefox menus, dialogs and the sidebar.

**Comment by samrocketman** [comment]
I use the following `.vimperatorrc` settings on a vanilla install:

```
; move left tabs with SHIFT+J
nmap J gT
; move right tabs with SHIFT+k
nmap K gt
; use a home row mapping for following links
set hintchars=hjklasdfgyuiopqwertnmzxcvb
```

Common operations for me include:

* Normal mode:
  - follow links - `f`, `F`
  - tabopen - `t`, `T`
  - open - `o`, `O`,
  - (undo) close tab - `d`, `u`
  - page marking and jumping - `m`, `'`
  - jump top/bottom - `g`, `G`
* Command mode:
  - tab buffers - `:b`, `:ls`
  - yank `<a>` names - `;#`
  - copy links instead of following them - `;y`

> **IMPORTANT** - The primary reason why I find vimperator useful over other clones (like vimium for Chrome) is the command mode bar.  Additionally, navigating tabs doesn't annoyingly force-focus the URL bar like it does in Chrome.  I hope going forward vimperator doesn't adopt this poor behavior.  I like using the ESC key.

If I think of more, I'll update my commit.

**Comment by nachoparker** [comment]
oh yeah, forgot two

y - copy the URL of current tab
CTRL+^ go to last tab. This one is awesome! ( or CTRL+6)

**Comment by rr-** [comment]
What I use, non-exhaustive list:

- <kbd>f</kbd> <kbd>F</kbd>
- <kbd>;</kbd><kbd>i</kbd> and similar
- <kbd>gg</kbd>, <kbd>G</kbd>, <kbd>50j</kbd>, etc.
- **<kbd>[[</kbd> and <kbd>]]</kbd>**, very underrated
- tabs (<kbd>gt</kbd>, <kbd>d</kbd>, <kbd>P</kbd>, `:xall`)
- system clipboard (<kbd>y</kbd>, <kbd>P</kbd>)
- history (<kbd>H</kbd>, <kbd>L</kbd>)
- bookmarks (<kbd>A</kbd>)
- zooming (<kbd>4zi</kbd>)
- keyboard shortcuts remapping (duh)
- GUI (hide everything but tabs and command line) (we'll probably have to pester Mozilla to include this natively)
- external form / textarea editing with `gvim`
- `:js` - open with mpv, including custom hint mode
- integration with `<input type=file>` (see my pull requests, btw) to avoid the hellish GTK file dialog
- `:set hintchars=123456`
- `:set browser.*` like this
    ```
    "fit image to window
    noremap zf :set! browser.enable_automatic_image_resizing=true<CR><Esc>:reload<CR>
    
    "show image at original size
    noremap zo :set! browser.enable_automatic_image_resizing=false<CR><Esc>:reload<CR>
    ```

**Comment by cmcaine** [comment]
Process spawning is fine: we just do the same procedure as https://github.com/vimperator/vimperator-labs/issues/729#issuecomment-283336371

We'd probably just define a "native application" that's a python script that just runs whatever we ask it to, and use that script for all external invocations.

**Comment by cmcaine** [comment]
`set browser.` doesn't do anything for me... Is it intended to make changes to the settings on about:config?

**Comment by rr-** [comment]
Yes, I meant the family of `browser.` options followed by anything browser-specific from `about:config`

**Comment by cmcaine** [comment]
Ah, you have to use `set!`, with the `!`. Cool: didn't know about that. Not sure we'll be able to reimplement, but thanks for the use case.

**Comment by marza91** [comment]
Don't think there's anything here I'm missing that hasn't already been mentioned, but the f and F keys always worked surprisingly well for me when switching from Vimium+Chrome to Vimperator+Firefox, just because it seems to highlight many links/buttons that Vimium couldn't *Ugly look at the MS Outlook Web App*

**Comment by rr-** [comment]
At very least, this issue is a good place to gather the feature set to put in such a survey.

**Comment by SammysHP** [comment]
- Commandline/statusbar
- <kbd>o</kbd>, <kbd>O</kbd>, <kbd>t</kbd>, <kbd>T</kbd>, <kbd>w</kbd>, <kbd>W</kbd> – they must open something like the completion in Vimperator.
- <kbd>f</kbd>, <kbd>F</kbd> and <kbd>;y</kbd> to follow and copy links
- <kbd>/</kbd> with the current behavior, the normal findbar doesn't work for me
- <kbd>y</kbd> and <kbd>p</kbd> to copy and paste URL
- Mappings that can override Firefox keybindings like <kbd>Ctrl</kbd>+<kbd>Q</kbd>
- <kbd>b</kbd> to show and switch buffers
- `:set gui=none` – just me, the content and a statusbar
- Some kind of NoScript integration. Showing the popup would be enough.
- Keyboard navigation of course (<kbd>h</kbd>, <kbd>j</kbd>, <kbd>k</kbd>, <kbd>l</kbd>, <kbd>gg</kbd>, <kbd>G</kbd>)
- Open an external editor (<kbd>Ctrl</kbd>+<kbd>I</kbd>) would be nice
- Commands to change preferences (`:set! …`)
- Tab commands (<kbd>d</kbd>, <kbd>u</kbd>, `:tabdetach`, `:tabmove`)
- Passthrough mode (<kbd>Ins</kbd>)

**Comment by protrolium** [comment]
everything @SammysHP mentions is essential to me, as well. I would add: 
`[[` and `]]` for pagination 
and `go*` + `gn*` for accessing qmarks.

**Comment by nachoparker** [comment]
I did not know ]] !!!!!

genius!!!!!!!!!!!

**Comment by MoSal** [comment]
There is also `C-a` and `C-x` for incrementing/decrementing the last number in URL.

**Comment by nachoparker** [comment]
> There is also C-a and C-x for incrementing/decrementing the last number in URL.
goddamn awesome!

I use C-6 or C-^ a lot to switch to the last used tab

**Comment by aplaice** [comment]
In addition to what has been already listed here — `:set gui=none`, `buffers!`, quickmarks and local marks, `]]`, obviously `f`/`F`, etc. — I really, really love the vim-like self-documentation `:help` + key or `:help` + command/topic. 

(I'd say that I'd have been even more ecstatic about an emacs-like `C-h k` (which has the two advantages that one can just press the keys exactly as one would use them (i.e. just press `C-h k C-a` on your keyboard, rather than type in `:help <C-a>`) and that it's updated to user-modifications, in case you forget what you added to your config a year ago), but I'd probably be lynched, it'd be a new feature request, and keeping the existing `:help` in the webextension vimperator would still be amazing.)

**Comment by cmcaine** [comment]
The Emacs-style help idea is interesting, and probably wouldn't be hard. Thanks for mentioning it.

**Comment by samrocketman** [comment]
Half page navigation `ctrl+u` and `ctrl+d` are additional shortcuts I enjoy using.

**Comment by tecfu** [comment]
I use caret mode / visual mode extensively for search and copy.

**Comment by patrick-ausderau** [comment]
In addition to what is already listed I frequently use `gu`

**Comment by nachoparker** [comment]
oh forgot `gu`.... don't use it much but I do use `gr` a LOT

**Comment by patrick-ausderau** [comment]
And I forgot `ctrl+w` in insert mode or when editing an URL (e.g. from a wikipedia page to another: `shift+O`, then `ctrl+w`, a letter or two and `tab` + `enter`)

**Comment by timss** [comment]
Nice list, but I'd personally bump `js` up to medium because it's so important for extensibility with other addons (TST, ..), user scripts and even running scripts/programs in your OS from FF<sup>1</sup>. Functions and commands using what is really `eval()` with sugar are something I use all the time.

I'd also like to echo the usefulness of `set!`. Not only can you configure most of FF through these preferences, but also many addons. Reinstalling or deploying my config to new profiles/installations isn't that tiresome much because of this.

<sup>1</sup> Altough I realise this doesn't seem like it's going to be possible in the same way as we have now in the era of WE anytime soon.

**Comment by Travolter** [comment]
Biggest, almost inevitable, loss for me will/would be the gui=None. I genuinely dislike having an url bar eat up vertical space.

**Comment by cmcaine** [comment]
@timss, interop with addons like TST will now rely on the active support of those addon developers (see below) and may be somewhat complicated as the addons will execute in different processes. `js` eval is not otherwise problematic.

@rr, @SammysHP, @timss, on `set!` word of mozilla is thus:

```
18:07 <@aswan> colinc: there won't be general access
18:07 <@aswan> we have apis like privacy, proxy, etc that give controlled access to individual prefs
18:07 <@aswan> and we can add more like those
```

It's also worth noting that webextensions will not be able to add keys to about:config, so no configurability of addons will be possible through that interface in the future. Interop with other addons will have to be done by either the addon community agreeing on some common API/configuration system or by pull requests on each project.

If anyone has the time to look into this, it could be an interesting project for Vimperatoralikes and the wider webextension community.

**Comment by fakedrake** [comment]
This thread is a goldmine of vimperator features I had no idea about. Quite ironic that they are the features in the greatest danger of being discontinued.

**Comment by cmcaine** [comment]
@Travolter, It will be possible to hide the tab bar, but Mozilla devs are strongly opposed to allowing the navigation bar to be hidden. See: https://bugzilla.mozilla.org/show_bug.cgi?id=1332447

I don't understand their security concerns, but they seem fairly set about it. At some point, I'd like to get some direct contact with someone from the security team to try and understand their position better.

If any of you do decide to post on bugzilla, please remember that the only productive way to engage is politely and constructively - webextensions, multiprocessing and the other breaking changes are, IMHO, good things for firefox. Too many people are already ranting unjustifiably at the Firefox team about this.

**Comment by Travolter** [comment]
@cmcaine I had read that bug, and I am aware of the problem. It's why I mentioned it, to highlight how important that feature is to some of us. I sincerely hope a solution can be found. As far as I am aware their security concerns have to do with the green SSL icon and probably url redirecting. Which both only make sense for the non tech savvy users, not for people who go around messing with their browser UI.

**Comment by dasJ** [comment]
I could live with a subset of the functionality proposed here. The question is: When do we start? Who is 'we'? Will there be a new repository in the `vimperator/` namespace or does somebody else take it?
By quickly reading over this thread, I found people saying there willing to contribute and I think the earlier the project gets started, the better.
Maybe WebExtensions doesn't support all features we need, but we should starting to implement the  functionality it already has.

So does anyone have more information on the near future? @gkatsev

**Comment by cmcaine** [comment]
See issue #705.

@bovine3dom and I are working on a replacement and some chrome addons are beimg ported.

For now, I am working with Mozilla to develop the new APIs we need for approximate feature parity with vimperator.

If you would like to help with our project, please mail me or open an issue on my tridactyl repo.

**Comment by fakedrake** [comment]
After switching to chrome for a short while because of my job, I had to use cvim. A huge problem with cvim (which is much closer to vimperator than vimium) is that it is based on injecting some code in each page. The problem with that is that a) not all pages are injectable (eg pdfs and chrome links) and b) no control until the page has sufficiently loaded.

How well is the webextension port doing in this respect?

**Comment by bovine3dom** [comment]
At the moment, it has the same problem. We would need a new WebExtension API to deal with it; we are currently waiting on Mozilla for feedback on our proposal. That's the Keyboard API that has been spoken about in this thread.                                                                                                                                                                                                                                                                                                                                                    From: notifications@github.comSent: 1 September 2017 10:05 amTo: vimperator-labs@noreply.github.comReply to: reply@reply.github.comCc: oliver.blanthorn@postgrad.manchester.ac.uk; mention@noreply.github.comSubject: Re: [vimperator/vimperator-labs] Feature wishlist for WebExtensions port (#729)  After switching to chrome for a short while because of my job, I had to use cvim. A huge problem with cvim (which is much closer to vimperator than vimium) is that it is based on injecting some code in each page. The problem with that is that a) not all pages are injectable (eg pdfs and chrome links) and b) no control until the page has sufficiently loaded.
How well is the webextension port doing in this respect?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by bovine3dom** [comment]
> Slightly tangential to this issue: what do people think about redirecting Vimperator users (or maybe a subset of users) to a survey of feature importance in some future update?

FF57 is going to be released on November 14th. I think it might be worth having a kind of "exit poll" - one last update to the AMO within the next weeks, directing people to a survey containing:

- a short note saying that Vimperator is dying and explaining why,
- a short list of alternatives (Qutebrowser et al),
- and a text area asking people to list their favourite features of Vimperator separated by newlines.

Hopefully such a survey would be useful to any successor projects, and also to the users themselves.

**Comment by timss** [comment]
Good idea, @bovine3dom 👍 

How about a Google Forms survey with public results? I'd be willing to make one, but it would be nice with some feedback/collaborators before it's published.

cc @maxauthority @gkatsev (who have access to AMO).

**Comment by cmcaine** [comment]
I'm happy to give feedback on it.

**Comment by maxauthority** [comment]
Yeah, having a good "final words" post when FF57 comes out sounds good, and
I'm certainly going to post it on AMo and vimperator.org. But input what
shall be written there is certainly helpful given the fact that I currently
have even less time unfortunately (became a dad 7 weeks ago).

Since vimperator.org's homepage is hosted on github, people could also send
me pull requests for good content.

Sorry for being so little responsive lately but great to see so many people
here still doing good stuff for the community!

--
Martin

**Comment by bovine3dom** [comment]
I'd be happy to give feedback / ideas.

For a start - I think most people have missed that using Firefox's "Extended Support Release" will let them use Vimperator until June 2018 [1], so it might be wise to mention that.

Other alternatives:
- Qutebrowser
- uzbl
- Vimium-FF [2]
- Tridactyl (maybe)
- Palemoon & Pentadactyl

[1] https://blog.mozilla.org/addons/2017/10/03/legacy-add-on-support-on-firefox-esr/
[2] https://addons.mozilla.org/en-GB/firefox/addon/vimium-ff/

**Comment by Koushien** [comment]
Don't have much else to add that others haven't (I'd be happy to contribute to the exit poll if desired), but

>became a dad 7 weeks ago

Congratulations @maxauthority!

**Comment by The-Compiler** [comment]
Some more alternatives other than what @bovine3dom mentioned:

- [luakit](https://luakit.github.io/) (recently gained a lot of traction again after making a WebKit2-fork the official "successor")
- [vimb](https://fanglingsu.github.io/vimb/) (quite minimal, and not much activity recently...)
- [Saka Key](https://key.saka.io/) (another Firefox WebExtension)

(disclaimer: I haven't used any of the alternatives for longer than a few minutes, with the exception of qutebrowser, which I'm the author of)

And congratulations @maxauthority on becoming a dad! Don't worry about not being responsive, family is more important than GitHub :wink:

**Comment by timss** [comment]
Congratulations @maxauthority ! :D 

---

I've created an initial version of the survey which I'd like feedback on/help completing. 

https://docs.google.com/forms/d/1WtS8eBXpLIyVXl2j937N-PP4nKX7ZjKL7P5cCEn2nc0/edit?usp=sharing

Feel free to add stuff or comments, especially where I've added TODOs. I'm not aware of any easy way to share edit permissions without anyone possibly tampering with it (now or later), but hopefully we'll be OK. Responses will be deleted before survey is published on AMO etc.

>(disclaimer: I haven't used any of the alternatives for longer than a few minutes, with the exception of qutebrowser, which I'm the author of)

Similarly, I don't have experience with all of these addons or browsers. Please keep in mind when reviewing the survey :)

**Comment by bovine3dom** [comment]
That's a good start, I think.

I don't quite see the point of section 3. Everyone viewing the survey will be using Firefox and Vimperator, so almost everyone will just answer that. I don't think it's a very interesting question.

I'd rather keep it to either just a free text area asking people what they liked in Vimperator, or that and a simple "Do you want a replacement for Vimperator?". Currently, there's a lot to scroll through.

**Comment by timss** [comment]
Thanks for having a look. I like your introduction better, and the colors (if that were you) :)

I was sort of thinking it could serve as a survey not only for current users of Vimperator, but also for those who've moved to other projects. Although, maybe it's better for this survey to be focused on informing about alternatives and gathering interesting user info (primarily what users like about Vimperator).

>[..] or that and a simple "Do you want a replacement for Vimperator?". 

Maybe this fits into section 4?

>Currently, there's a lot to scroll through.

Fair enough, was thinking the same! Although, sections will be paginated I believe.

**Comment by aplaice** [comment]
I've made some tiny stylistic changes.

Regarding the "favourite features" section, I think that in addition to the lists listed in this issue, we could look at the feature list from `:helpall` (copied verbatim below). This would have the advantage that our biases wouldn't result in some features being missed from the survey.

    * Vim-like keybindings (h, j, k, l, gg, G, 0, $, ZZ, <C-f>, etc.)

    * Ex commands (:quit, :open www.foo.com, …) with a proper command line

    * Tab completion available for all commands, showing suggestions while you type

    * Hint mode (start with f to follow a link)

    * Extensions! Yes, you can extend Vimperator's functionality with scripts just like you can extend Firefox with extensions

    * Explore JavaScript objects with :echo window and even context-sensitive tab completion

    * Easily customizable GUI (easily hide all GUI elements with :set gui=none)

    * Ability to :source JavaScript files, and to use a ~/.vimperatorrc file with syntax highlighting if you install vimperator.vim

    * Easy quick searches (:open foo will search for "foo" in google, :open ebay terminator will search for "terminator" on ebay) with support for Firefox keyword bookmarks and search engines

    * Count supported for many commands (3<C-o> will go back 3 pages)

    * Beep on errors

    * Marks support (ma to set mark a on a web page, 'a to go there)

    * QuickMarks support (quickly go to previously marked web pages with go{a-zA-Z0-9})

    * :map and :command support (and feedkeys() for script writers)

    * :time support for profiling

    * Move the text cursor and select text with Vim keys and a Visual mode

    * External editor support

    * Macros to replay key strokes

    * AutoCommands to execute actions on certain events

    * A comprehensive help system, explaining all commands, mappings and options

I think that the main points present in the feature list but not mentioned in the GitHub issue are `:time` and `AutoCommands` (I had not been aware of these, but it's possible that they have a dedicated following somewhere on the internet).

**Comment by bovine3dom** [comment]
@aplaice there are so many features that I think just asking people to write their own in is more sensible. Just scroll up through this thread to see what diversity there is.

I think it is important to keep it as short and simple as possible. Potentially, it might be worth having three or four main options for alternatives, and then to provide a link to a related projects page with more detail.

**Comment by klas-genestack** [comment]
> there are so many features that I think just asking people to write their own in is more sensible. Just scroll up through this thread to see what diversity there is

It's easy to forget some feature you use.  Maybe it's worth adding a link to this thread?  I learnt a bunch of nice features myself from it.

**Comment by krigstask** [comment]
> there are so many features that I think just asking people to write their own in is more sensible. Just scroll up through this thread to see what diversity there is

It's easy to forget some feature you use.  Maybe it's worth adding a link to this thread?  I learnt a bunch of nice features myself from it.

**Comment by samrocketman** [comment]
I've tweeted a sad message to `@mozilla` and `@firefox` that vimium is basically the only reason I'm using Firefox.  I've now got to switch to Chrome because Vimium is the second most comfortable extension I use (not as comfortable or featured as vimperator).  https://twitter.com/sag47/status/917434763001999360

**Comment by Croydon** [comment]
@samrocketman I still don't get this argument. You are saying you are switching to a browser (Chrome) with even less opportunities for add-ons (i.e. less powerful add-ons), because the new Firefox has less powerful add-ons APIs than the old Firefox, but still more than Chrome... and growing.

**Comment by bovine3dom** [comment]
@samrocketman just use Firefox ESR. You're safe until June then.

**Comment by samrocketman** [comment]
@Croydon I'm not really asking you to understand my argument.  My statement is fairly simple and a preference for me.  I don't really want to hijack this issue to convince you of my point of view.  We can chat on twitter if you'd like to continue the conversation :wink: I'm `@sag47`.

In general, if volunteer development is not enough (or currently available Firefox APIs are not enough), then hopefully Mozilla/Firefox will listen and provide the necessary APIs or development support necessary for vimperator to continue.

@bovine3dom I'll search my package repository for Firefox ESR.  That's a good idea :smile: if it allows me to use it a little longer :heart:

**Comment by cmcaine** [comment]
I think the purpose of the survey should be:

 1. To inform users that vimperator is dying
 2. Inform them of alternatives
 3. Inform us about what features people like and how important they are to them

I think 2 should be mostly done via a section in the README on this project. That makes the content more accessible and discoverable, means we can use proper markup and shortens the exit poll.

I suggest that the form should be:

 1. Intro and goodbye
 2. Brief survey of vimperator features
 3. Link to a markdown page/wiki page in this repo talking about alternatives

**Comment by timss** [comment]
> It's easy to forget some feature you use. Maybe it's worth adding a link to this thread? I learned a bunch of nice features myself from it.

> Also, it seems a bit silly to ask people to put a wishlist in the poll (with results opaque to the users) when this GitHub issue exists. You should just reference it if you still want to do the exit poll.

It was already referenced in the introduction, but I've added another link to this issue in the section in question (4).

> I think 2 should be mostly done via a section in the README on this project. That makes the content more accessible and discoverable, means we can use proper markup and shortens the exit poll.

Good idea. Similar lists can be found in your tridactyl repo, or qutebrower's, both of which have been helpful. The one in Vimperator should be fairly short and unopinionated though, just as was intended in the survey.

I could try to look into it when I got some time on my hands, but until then feel free to create a PR. Take whatever is included in the alternatives section of the survey now deemed useful.

---

> @bovine3dom I'll search my package repository for Firefox ESR. That's a good idea 😄 if it allows me to use vimperator a little longer ❤️

It's available in some distros (Debian, but only PPA on Ubuntu), but Firefox is fully self contained and is therefore also easy to install manually. Personally I have a few installations laying around in `/opt/firefox` with some symlinks for latest esr etc.

**Comment by cmcaine** [comment]
See PR #800.

**Comment by timss** [comment]
Merged and linked to in the survey. The lack of formatting options in Google Forms is kind of annoying, so it might be better to just make it a seperate section for some boldness and stuff.

Since alternatives are gone, I opted for removing the section about which current browsers and if any, which Vim- or Vimperator-like addons/extensions people are currently using. By now it felt out of place.

As for the last free text section where we ask to list favorite features, someone filled in and replaced the question title with his/her answer ("Keyboard shortcuts, custom commands, GUI modifications, external commands execution"), lol. I tried remembering what it used to say, but feel free to modify however you see fit, @cmcaine, @bovine3dom or others. And thanks for the help so far! :)

**Comment by MoSal** [comment]
> It's available in some distros (Debian, but only PPA on Ubuntu), but Firefox is fully self contained and is therefore also easy to install manually. Personally I have a few installations laying around in /opt/firefox with some symlinks for latest esr etc.

@timss  @bovine3dom It should be noted that profiles are no longer backwards-compatible (since FF55, I think). So, you can't go back from FF56 to ESR with the same profile.

**Comment by timss** [comment]
Well that's nice to know..

**Comment by laurentbardin** [comment]
@MoSal I've noticed no issues when I started using ESR this week, on macOS, Linux, and Windows. All my windows, tabs, addons, and preferences were correctly restored as far as I could tell. Any specific problem I should be expecting?

**Comment by cmcaine** [comment]
History autocompletion didn't work for me.

**Comment by MoSal** [comment]
@laurentbardin 

I wouldn't know what to expect.

> Firefox does not support downgrades, even though this may have worked in past versions. Users who install Firefox 55+ and later downgrade to an earlier version may experience issues with Firefox.

Reading this [in the release notes](https://www.mozilla.org/en-US/firefox/55.0/releasenotes/) deterred me enough to not even try.

**Comment by laurentbardin** [comment]
@MoSal Thanks for the heads up, I missed that changelog item for some reason. So far, so good, fingers crossed.

**Comment by timss** [comment]
FYI details regarding the "EOL" of Vimperator as well as the survey is now up on vimperator.org and AMO. Thanks @maxauthority and @gkatsev 

- https://github.com/vimperator/vimperator.github.io/issues/8
- http://vimperator.org/vimperator
- http://vimperator.org/about
- https://addons.mozilla.org/en-US/firefox/addon/vimperator/

I've locked the survey and it's now available on the following URLs:

- [survey](https://goo.gl/forms/TFfLBxpbjNJfd7Gg2)
- [results](https://docs.google.com/forms/d/e/1FAIpQLSfTvWhTWbWAWpbbeoVM_RkDEQV6dqp4YL9Gt77gNkQHN4pTZA/viewanalytics)

**Comment by bovine3dom** [comment]
That's great. Thanks to all involved.

Is there any intention to update Vimperator to have a pop up with the same information so that current users can be informed?

Just a new tab that shows up once with the survey would be enough I think.

**Comment by timss** [comment]
And thanks to you and the others here who helped out with the survey, documentation of alternatives and of course a special thanks to those who are currently working with alternatives to Vimperator- or vim-like browser or addon alternatives! 👍 Personally I'll stick to ESR for a while longer, it's going to be hard to change even though the work looks promising.

There are as far as I know no plans for something like that, no, but it could probably be done using the same piece of code responsible for opening `:help` the first time:

https://github.com/vimperator/vimperator-labs/blob/ae0df8040bc2821d6d0ce39a9c2f3b8feda4a3e4/common/content/liberator.js#L1834

I'm looking into creating something.
